util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+        part_offset = self.config['partitions'][part_idx]['offset']
+        # First chop up all items into individual bytes.
+        data_bytes = [0] * part_size
+        # Annotation is propagated into the hexfile as comments
+        annotation = ['unallocated'] * part_size
+        # Need to keep track of defined items for the scrambling.
+        # Undefined regions are left blank (0x0) in the memory.
+        defined = [False] * part_size
+        for item in self.config['partitions'][part_idx]['items']:
+            for k in range(item['size']):
+                idx = item['offset'] - part_offset + k
+                annotation[idx] = part_name + ': ' + item['name']
+                if 'value' in item:
+                    data_bytes[idx] = ((item['value'] >> (8 * k)) & 0xFF)
+                    defined[idx] = True
+
+        # Reshape this into 64bit blocks (this must be aligned at this point)
+        if len(data_bytes) % 8:
+            raise RuntimeError('Partitions must be 64bit aligned')
+
+        data_blocks = []
+        data_block_defined = []
+        for k, b in enumerate(data_bytes):
+            if (k % 8) == 0:
+                data_blocks.append(b)
+                data_block_defined.append(defined[k])
+            else:
+                data_blocks[k // 8] += (b << 8 * (k % 8))
+                # If any of the individual bytes are defined, the
+                # whole block is considered defined.
+                data_block_defined[k // 8] |= defined[k]
+
+        # Check if scrambling is needed
+        if self.config['partitions'][part_idx]['secret']:
+            part_name = self.config['partitions'][part_idx]['name']
+            key_sel = self.config['partitions'][part_idx]['key_sel']
+            log.info('> Scramble partition with key "{}"'.format(key_sel))
+
+            for key in self.config['scrambling']['keys']:
+                if key['name'] == key_sel:
+                    break
+            else:
+                raise RuntimeError(
+                    'Scrambling key cannot be found {}'.format(key_sel))
+
+            for k in range(len(data_blocks)):
+                if data_block_defined[k]:
+                    data_blocks[k] = _present_64bit_encrypt(
+                        data_blocks[k], key['value'])
+
+        # Check if digest calculation is needed
+        if self.config['partitions'][part_idx]['hw_digest']:
+            # Digest is stored in last block of a partition
+            if self.config['partitions'][part_idx].setdefault('lock', False):
+                log.info('> Lock partiton by computing digest')
+                # Digest constants at index 0 are used to compute the
+                # consistency digest
+                iv = self.config['scrambling']['digests'][0]['iv_value']
+                const = self.config['scrambling']['digests'][0]['cnst_value']
+                data_blocks[-1] = _present_64bit_digest(
+                    data_blocks[0:-1], iv, const)
```

Comment:
Does this need a check that we don't have a value in `data_blocks[-1]` already?

Reply:
Yeah you are right that we should check and this explicitly and disallow adding a digest in case of HW-governed digests. I'll add a check and an exception.